lib/spack/spack/environment/environment.py

```diff
@@ -556,9 +556,15 @@ def _is_dev_spec_and_has_changed(spec):
         # Not installed -> nothing to compare against
         return False
 
-    _, record = spack.store.STORE.db.query_by_spec_hash(spec.dag_hash())
-    mtime = fs.last_modification_time_recursive(dev_path_var.value)
-    return mtime > record.installation_time
+    # hook so packages can use to write their own method for checking the dev_path
+    # use package so attributes about concretization such as variant state can be
+    # utilized
+    if hasattr(spec.package, "detect_dev_src_change"):
+        return spec.package.detect_dev_src_change()
+    else:
+        _, record = spack.store.STORE.db.query_by_spec_hash(spec.dag_hash())
+        mtime = fs.last_modification_time_recursive(dev_path_var.value)
+        return mtime > record.installation_time
```

Comment:
Wouldn't it be cleaner to have this be simply

```suggestion
    # hook so packages can use to write their own method for checking the dev_path
    # relevant for packages with sparse checkouts primarily
    return spec.package.detect_dev_src_change()
```
and then have a method on `PackageBase` that sets the default:
```
    def detect_dev_src_change(self):
        _, record = spack.store.STORE.db.query_by_spec_hash(self.spec.dag_hash())
        mtime = fs.last_modification_time_recursive(dev_path_var.value)
        return mtime > record.installation_time

Reply:
yeah I thought about this, but I went this way to admittedly short cut potential impacts on test infrastructure. I'm fine with doing this if you'd prefer but I'm trying to get this in ASAP since I'm in an iterate and develop mode right now for the sparse path paradigm